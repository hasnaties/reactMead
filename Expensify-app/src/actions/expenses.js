import database from "../firebase/firebase";

// ADD_EXPENSE Action OBJ
export const addExpense = (expense) => ({
  type: "ADD_EXPENSE",
  expense
});

export const startAddExpense = (expenseData = {}) => {
  
  return (dispatch) => {
    const {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
    } = expenseData;

    const expense = {description, note, amount, createdAt}
    return database.ref('expenses').push(expense).then((ref) => {
      dispatch(addExpense({
        id: ref.key,
        ...expense
      }));
    }).catch((err) => {
      console.log(err);
    });
  };
};

// EDIT_EXPENSE Action OBJ
export const editExpense = (id, updates) => ({

  type: "EDIT_EXPENSE",
  id,
  updates
});

// remove Expense Action OBJ
export const removeExpense = ({ id } = {}) => ({

  type: "REMOVE_EXPENSE",
  id
});

// set Expenses Action OBJ
export const setExpenses = (expenses) => ({

  type: "SET_EXPENSES",
  expenses
});

export const startSetExpenses = () => {

  return (dispatch) => {
    return database.ref('expenses')
      .once('value')
        .then((snapshot) => {
          const expensesData = [];

          snapshot.forEach((childSnap) => {
            expensesData.push({
              id: childSnap.key,
              ...childSnap.val()
            });
          });

          dispatch(setExpenses(expensesData));
        });
  };
};