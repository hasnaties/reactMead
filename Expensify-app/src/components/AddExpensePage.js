import React from 'react';
import ExpenseForm from './ExpenseForm';

const AddExpensePage = () => (
  <div>
    <h4>Add Expense</h4>
    <ExpenseForm />
  </div>
);

export default AddExpensePage;