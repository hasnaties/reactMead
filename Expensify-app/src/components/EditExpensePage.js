import React from 'react';

const EditExpensePage = () => (
  <div>
    <p>Edit</p>
  </div>
);

export default EditExpensePage;