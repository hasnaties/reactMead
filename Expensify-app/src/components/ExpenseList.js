import React from 'react';
import { connect } from 'react-redux';
import selectExpenses from '../selectors/expenses';
import ExpenseListItem from './ExpenseListItem';

export const ExpenseList = (props) => (
  
  <div>
    {props.expenses.length === 0 ? (
      <p>Expenses list is empty.</p>
    ) : (
      <div>
        <h4>Expenses List</h4>
        {
          props.expenses.map((expense) => {
          return (
            <ExpenseListItem key={expense.id} {...expense}/>
          );
          })
        }
      </div>
    )}
  </div>
);

const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  }
}

export default connect(mapStateToProps)(ExpenseList)