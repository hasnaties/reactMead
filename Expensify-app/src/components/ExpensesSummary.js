import React from 'react';
import { connect } from "react-redux";
import selectExpenses from '../selectors/expenses';
import expensesTotal from '../selectors/expenses-total';
import numeral from "numeral";

export const ExpensesSummary = (props) => (
  <div>
    <p>
      Viewing {props.expenseCount} expense{props.expenseCount > 1 ? 's' : undefined} totalling {numeral(props.expensesTotal / 100).format('$0,0.00')}
    </p>
  </div>
);

const mapStateToProps = (state) => ({
  expenseCount: selectExpenses(state.expenses, state.filters).length,
  expensesTotal: expensesTotal(selectExpenses(state.expenses, state.filters))
});

export default connect(mapStateToProps)(ExpensesSummary);