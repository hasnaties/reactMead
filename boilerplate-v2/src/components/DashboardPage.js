import React from 'react';

const ExpenseDashboardPage = () => (
  <div>
    Dashboard content.
  </div>
);

export default ExpenseDashboardPage;