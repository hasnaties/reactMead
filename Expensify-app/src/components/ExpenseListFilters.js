import React from 'react';
import { connect } from 'react-redux';
import { editTextFilter, setEndDate, setStartDate, sortByAmount, sortByDate } from '../actions/filters';
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
  
  state = {
    calendarFocused: null
  }

  onDatesChange = ({startDate, endDate}) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  }

  onFocusChange = (calendarFocused) => {
    this.setState({ calendarFocused })
  }

  onTextChange = (e) => {
    return (this.props.editTextFilter(e.target.value));
  }

  onSortByChange = (e) => {
    if(e.target.value === "date") {
      return (this.props.sortByDate());
    }
    return (this.props.sortByAmount());
  }
  
  render () {
    return (
      <div>
        <input type='text' 
        value={this.props.filters.text} 
        onChange={this.onTextChange}>
        </input>
    
        <select 
        value={this.props.filters.sortBy} 
        onChange={this.onSortByChange}
        >
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>
        
        <DateRangePicker
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
          showClearDates={true}
        />
      </div>
    )
  }
}

const mapDispatchToProps = (dispatch) => ({
  
  setStartDate: (startDate) => dispatch(setStartDate(startDate)),
  setEndDate: (endDate) => dispatch(setEndDate(endDate)),
  editTextFilter: (value) => dispatch(editTextFilter(value)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount())
});

const mapStateToProps = (props) => ({
  filters: props.filters
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters)