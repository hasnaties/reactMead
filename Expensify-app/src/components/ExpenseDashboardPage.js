import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => (
  <div>
    <p>Dashboard.</p>
    <ExpensesSummary />
    <ExpenseListFilters />
    <ExpenseList />
  </div>
);

export default ExpenseDashboardPage;