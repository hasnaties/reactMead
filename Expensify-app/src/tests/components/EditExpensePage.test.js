import React from 'react';
import { EditExpensePage } from '../../components/EditExpensePage';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';

let startEditExpense, startRemoveExpense, history, wrapper;

beforeEach(() => {
  startEditExpense = jest.fn();
  startRemoveExpense = jest.fn();
  history = {
    push: jest.fn()
  };
  wrapper = shallow(
    <EditExpensePage 
      startEditExpense={startEditExpense} 
      startRemoveExpense={startRemoveExpense}
      history={history}
      expense={expenses[0]}
    />
  )
});

test('should Render EditExpensePage correctly', () => {
  
  expect(wrapper).toMatchSnapshot();
});

test('should Handle editExpense', () => {
  
  wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
  expect(history.push).toHaveBeenLastCalledWith("/");
  expect(startEditExpense).toHaveBeenLastCalledWith(expenses[0].id, expenses[0]);
});

test('should Handle removeExpense', () => {
  
  wrapper.find('button').simulate('click');
  expect(startRemoveExpense).toHaveBeenLastCalledWith({id: expenses[0].id});
  expect(history.push).toHaveBeenLastCalledWith("/");
});