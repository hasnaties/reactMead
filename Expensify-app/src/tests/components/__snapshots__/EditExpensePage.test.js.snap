// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should Render EditExpensePage correctly 1`] = `
<div>
  <div
    className="page-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="page-header__title"
      >
        Edit Expense
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ExpenseForm
      expense={
        Object {
          "amount": 195,
          "createdAt": 0,
          "description": "this is the first test description.",
          "id": "1",
          "note": "this is note first.",
        }
      }
      onSubmit={[Function]}
    />
    <button
      className="button button--secondary"
      onClick={[Function]}
    >
      Remove Expense
    </button>
  </div>
</div>
`;
